src/sketch.lisp

```diff
@@ -182,136 +182,126 @@ used for drawing, 60fps.")
+			    :value value
+			    :defaultp defaultp
+			    :initarg initarg
+			    :accessor accessor)))
+
+(defun make-binding-from-default-slot (slot)
+  (let ((name (car slot))
+	(value (cadr (member :initform slot)))
+	(defaultp t)
+	(accessor (cadr (member :accessor slot)))
+	(initarg (cadr (member :initarg slot))))
+    (make-instance 'binding :sketch-name 'sketch
+			    :name name
+			    :channelp nil
+			    :original-value value
+			    :value value
+			    :defaultp defaultp
+			    :initarg initarg
+			    :accessor accessor)))
+
+(defmethod binding->slot ((binding binding))
+  `(,(binding-name binding)
+    :initarg ,(alexandria:make-keyword (binding-name binding))
+    :accessor ,(binding-accessor binding)))
+
+ ;;; DEFSKETCH macro
+
+(defun define-sketch-defclass (sketch-name custom-bindings)
+  `(defclass ,sketch-name (sketch)
+     ,(mapcar #'binding->slot custom-bindings)))
+
+
+(defun define-channel-observers (all-bindings)
+  (let ((channel-bindings (remove-if-not #'binding-channel-p all-bindings)))
+    (loop for binding in channel-bindings
+	  collect
+	  `(define-channel-observer
+	     (let ((win (kit.sdl2:last-window)))
+	       (when win
+		 (setf (,(binding-accessor binding) win) ,(binding-original-value binding))))))))
+
+  (defun define-draw-method (sketch-name all-bindings body)
+    `(defmethod draw ((*sketch* ,sketch-name) &key &allow-other-keys)
+       (with-accessors ,(mapcar (lambda (binding)
+				  (list (binding-name binding) (binding-accessor binding)))
+			 all-bindings)
+	   *sketch*
+	 ,@body)))
+
+(defun define-prepare-method (sketch-name default-bindings defined-bindings custom-bindings all-bindings)
+  (flet ((binding-in (bindings)
+	   (lambda (binding) (member binding bindings :test #'binding=))))
+    (let ((changed-bindings (remove-if-not (binding-in default-bindings) defined-bindings))
+	  (unchanged-bindings (remove-if (binding-in defined-bindings) default-bindings)))
+      `(defmethod prepare ((*sketch* ,sketch-name) &rest initargs &key &allow-other-keys)
+	 (declare (ignorable initargs))
+	 (let* (,@(loop for b in unchanged-bindings
+			collect `(,(binding-name b) (slot-value *sketch* ',(binding-name b))))
+		; We need to use the order from defined-bindings to support let*-like behavior
+		,@(loop for b in defined-bindings
+			when (funcall (binding-in changed-bindings) b)
+			  collect `(,(binding-name b) (if (getf initargs ,(binding-initarg b))
+							  (slot-value *sketch* ',(binding-name b))
+							  ,(binding-value b)))
+			when (funcall (binding-in custom-bindings) b)
+			  collect `(,(binding-name b) (or (getf initargs ,(binding-initarg b))
+							  ,(binding-value b)))))
+	   (declare (ignorable ,@(mapcar #'binding-name all-bindings)))
+	   (setf ,@(loop for b in all-bindings
+			 collect `(,(binding-accessor b) *sketch*)
+			 collect (binding-name b))))
+	 (setf (env-y-axis-sgn (slot-value *sketch* '%env))
+	       (if (eq (slot-value *sketch* 'y-axis) :down) +1 -1))))))
```

Comment:
This `setf` actually is not needed, as it should be handled by `(setf (sketch-x-axis ...) ...)`